[optests] Changed failures_dict format to json; automatic update of failures_dict

zou3519 opened the pull request:
We changed the failures_dict format from .py to json and added a way to
automatically update the failures dict (the user can set
PYTORCH_OPCHECK_ACCEPT=1 to do so), assuming the tests don't crash in the
process.

Some details:
- We introduced a FailuresDict class that handles save/load and from which one
can query a test status ("xfail", "skip", etc).
- PYTORCH_OPCHECK_ACCEPT=1 does not override everything. In particular: it
doesn't try to update the failures dict for a test marked as "skip", but it
will update it for tests marked as "xfail" or "success".
- PYTORCH_OPCHECK_ACCEPT=1 also does not override the "comment" field, unless
it is flipping an "xfail" into "success".
- I'll update the gdoc linked in the comments with how to actually use
PYTORCH_OPCHECK_ACCEPT=1 internally (it's not trivial).

Note that this isn't multithreading-safe, the current recommendation is to run
the tests sequentially if the user wants to use PYTORCH_OPCHECK_ACCEPT=1.

Differential Revision: D49167181

ezyang commented on test/minioptest_failures_dict.json:
```diff
@@ -0,0 +1,44 @@
+      }
+    }
+  }
+}
```
would be nice to do the trailing newline correctly

zou3519 replied:
yup

ezyang commented on test/test_custom_ops.py:
```diff
@@ -1803,9 +1785,11 @@ def test_delayed_error_no_requires_grad(self):
-    mini_op_test_failures_dict,
-    "test/test_custom_ops.py",
+    ["aten", "mini_op_test"],
+    get_file_path_2(
```
what's this helper function 🤔

zou3519 replied:
Something about internal fbcode having different codepaths to the resource :/

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -493,10 +548,63 @@ def resolve_unique_overload_or_throw(op: torch._ops.OpOverloadPacket):
+
+
+class FailuresDict:
+    def __init__(self, path, data):
```
type annotations would be nice

zou3519 replied:
sgtm

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -493,10 +548,63 @@ def resolve_unique_overload_or_throw(op: torch._ops.OpOverloadPacket):
+
+    def get_status(self, qualname, test_name):
+        if qualname not in self.data:
+            return "success"
```
interesting...

ezyang replied:
I guess this feels less bad when you realize this is "expect success", as opposed to "yolo it succeeded"

zou3519 replied:
My mental model is that we're talking about the status of the test:
- "skip" means we should skip the test
- "xfail" means we expect the test to fail
- "success" means we expect the test to succeed.
I could rename this "xsuccess" if we want for consistency (do you have other thoughts?) Though I'll probably land this first to avoid merge conflicts

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -493,10 +548,63 @@ def resolve_unique_overload_or_throw(op: torch._ops.OpOverloadPacket):
+
+        if status == "success":
+            # The default status is "success".
+            del dct[test_name]
```
lol

ezyang commented:
properly reviewed now 👍

zou3519 commented:
Thanks Ed. I need to rebase this anyways because the diff is complaining, so, will incorporate your comments

zou3519 committed:
[optests] Changed failures_dict format to json; automatic update of failures_dict (#109110)

Summary:

We changed the failures_dict format from .py to json and added a way to
automatically update the failures dict (the user can set
PYTORCH_OPCHECK_ACCEPT=1 to do so), assuming the tests don't crash in the
process.

Some details:
- We introduced a FailuresDict class that handles save/load and from which one
  can query a test status ("xfail", "skip", etc).
- PYTORCH_OPCHECK_ACCEPT=1 does not override everything. In particular: it
  doesn't try to update the failures dict for a test marked as "skip", but it
  will update it for tests marked as "xfail" or "success".
- PYTORCH_OPCHECK_ACCEPT=1 also does not override the "comment" field, unless
  it is flipping an "xfail" into "success".
- I'll update the gdoc linked in the comments with how to actually use
  `PYTORCH_OPCHECK_ACCEPT=1` internally (it's not trivial).

Note that this isn't multithreading-safe, the current recommendation is to run
the tests sequentially if the user wants to use PYTORCH_OPCHECK_ACCEPT=1.

bypass-github-export-checks

Test Plan: Existing tests

Reviewed By: ezyang

Differential Revision: D49167181
